### PR TITLE
Ports: Introduce support for Git repositories

### PR DESCRIPTION
**Pull request by timschumi**
This also includes a slightly updated `files` syntax, and slight cleanups in `./package.sh dev`.

### CODE REVIEW

File: Ports/.port_include.sh [hunk 1]
```diff
@@ -394,16 +437,30 @@ func_defined pre_patch || pre_patch() {
...
+                run git am --keep-cr --keep-non-patch "${filepath}"
+            else
                 run patch -p"$patchlevel" < "$filepath"
                 run touch .${filename}_applied
             fi
         done
     fi
+
+    if [ -e "${workdir}/.git" ]; then
+        run git tag -f original
```
Comment (gmta) [inline Ports/.port_include.sh:461]: Why is `original` _after_ patching? Maybe we should use `original` (or `upstream`, etc.) and `patched` as tags to differentiate between the revisions?

File: Ports/.port_include.sh [hunk 2]
```diff
@@ -366,6 +370,38 @@ fetch_simple() {
...
+    revision="${2}"
+
+    directory="$(basename "${repository}")"
+    backing_copy="${PORT_META_DIR}/${directory}"
+    working_copy="${PORT_BUILD_DIR}/${directory}"
+
+    run_nocd git init --bare "${backing_copy}"
+    run_nocd git -C "${backing_copy}" config core.autocrlf false
+    run_nocd git -C "${backing_copy}" worktree prune
+    run_nocd git -C "${backing_copy}" fetch --tags "${repository}" "${revision}"
```
Comment (gmta) [inline Ports/.port_include.sh:384]: For a standard (non-dev) checkout of a git repository, we can probably supply `--depth=1` here on initial fetch to minimize bandwidth and storage?

File: Ports/.port_include.sh (hunk 1 above)
Comment (timschumi) [reply Ports/.port_include.sh:461]: That naming is unchanged from `./package.sh dev`, where we had/have "import[ed tarball]", "original [state]" (i.e. before the user started `dev`) and the current state (which is wherever `HEAD` happens to be at).

I'll try and think of some better names.

File: Ports/.port_include.sh (hunk 2 above)
Comment (timschumi) [reply Ports/.port_include.sh:384]: We could, but not all ports (currently 0%; sample size 1) play nicely with that. For example, `gn` expects that, if you are building from inside Git, all the history down to the `initial-commit` tag is available.

At least `git fetch --unshallow` is a thing now, so if we decide to do it, it's going to be relatively painless.

File: Ports/.port_include.sh (hunk 1 above)
Comment (gmta) [reply Ports/.port_include.sh:461]: Ah then I mainly disagree with the original choices, let's see if we can come up with something in this PR. Maybe `source` (as downloaded from `$files`) and `patched` (state after patches)?

File: Ports/.port_include.sh (hunk 2 above)
Comment (gmta) [reply Ports/.port_include.sh:384]: Let's not then :^)

**Comment by EWouters** [comment]
I tried `./package fetch` on the gn port and it hangs the shell on this command:
```console
[gn/fetch] + git -C /Users/anon/git/serenity/Build/x86_64/Ports/gn/gn tag -f import fae280eabe5d31accc53100137459ece19a7a295 (nocd)
```
It hangs on trying to start my editor to input a message for the imported tag. Usually when I use `./package dev` I also have to manually add a message for the tags. I get dropped into an editor twice during the patch import process.
```console
File: /Users/erik/git/serenity/Ports/gn/gn/worktrees/gn/TAG_EDITMSG


#
# Write a message for tag:
#   import
# Lines starting with '#' will be ignored.
```
I thought this was the same for everybody, but it might be a difference in default git behavior instead, or settings. Adding `-m ''` to the git tag import commands on line 400 and 810 fixes that for me, but it might also suppress existing tag messages?

**Comment by alimpfard** [comment]
@EWouters afaik making annotated tags requires passing `-a` (or some other args I can't remember...) to `git tag`, do you have that as an alias or something? :thinking: 
passing `-m ''` will imply `-a` and make an annotated tag, not that it matters much

File: Ports/.port_include.sh [hunk 3]
```diff
@@ -747,64 +747,13 @@ do_dev() {
...
-                    else
-                        # The patch didn't apply, oh no!
-                        # Ask the user to figure it out :shrug:
-                        git am --keep-cr --keep-non-patch --3way "$patch" || true
-                        >&2 echo "- This patch does not apply, you'll be dropped into a shell to investigate and fix this, quit the shell when the problem is resolved."
-                        >&2 echo "Note that the patch needs to be committed into the current repository!"
-                        launch_user_shell
-                        force_patch_regeneration='true'
-                    fi
+                if ! git am --keep-cr --keep-non-patch --3way "$patch"; then
```
Comment (alimpfard) [inline Ports/.port_include.sh:815]: :yakkiss:

Comment (alimpfard) [reply Ports/.port_include.sh:815]: thanks github, I totally meant to submit this as a review 10/10 "cancel" feature.

**Comment by EWouters** [comment]
It was a setting after all, I had:
```console
[tag]
	gpgSign = true
```
in my global git config, which apparently also requires annotated tags. So as long as we don't care about signed tags it should not be a problem.

**Comment by alimpfard** [comment]
Then let's pass `--no-sign` to `git tag` here, that should fix it :P

**Comment by AtkinsSJ** [comment]
:warning: Conflict!

**Comment by EWouters** [comment]
Works great now, regardless of the global git setting.

**Review by ADKaster — APPROVED** [review]
Looks all reasonable to me :D

**Pull request merged**